build/RepositoryBuild.targets

```diff
@@ -0,0 +1,69 @@
+<Project>
+  <Import Project="$(_BuildGraphFile)" />
+
+  <Target Name="BuildRepositories">
+    <ItemGroup>
+      <BatchedRepository Include="$(MSBuildProjectFullPath)">
+        <AdditionalProperties>
+          RepositoryToBuild=%(RepositoryToBuildInOrder.Identity);
+          BuildRepositoryRoot=$(_CloneRepositoryRoot)%(RepositoryToBuildInOrder.Identity)\
+        </AdditionalProperties>
+        <BuildGroup>%(RepositoryToBuildInOrder.Order)</BuildGroup>
+      </BatchedRepository>
+    </ItemGroup>
+
+    <PropertyGroup>
+      <BatchBuilds Condition="$(BuildInParallel) AND '$(CompileOnly)'=='true'">true</BatchBuilds>
+      <BatchBuilds Condition="'$(BatchBuilds)'==''">false</BatchBuilds>
+    </PropertyGroup>
+
+    <MSBuild
+      Projects="@(BatchedRepository)"
+      BuildInParallel="$(BatchBuilds)"
+      Targets="_BuildRepository"
+      Properties="BuildGroup=%(BatchedRepository.BuildGroup)" />
+  </Target>
+
+  <Target Name="_BuildRepository" DependsOnTargets="_PinVersions">
+    <PropertyGroup>
+      <BuildScriptToExecute Condition="'$(OS)'!='Windows_NT'">build.sh</BuildScriptToExecute>
+      <BuildScriptToExecute Condition="'$(OS)'=='Windows_NT'">build.cmd</BuildScriptToExecute>
+      <RepositoryArtifactsDirectory>$(BuildRepositoryRoot)artifacts\build\</RepositoryArtifactsDirectory>
+    </PropertyGroup>
+
+    <Message Text="============ Building $(RepositoryToBuild) ============" Importance="High" />
+    <Exec
+      Command="./$(BuildScriptToExecute) $(_RepositoryBuildTargets) /p:BuildNumber=$(BuildNumber) /p:Configuration=$(Configuration)"
+      IgnoreStandardErrorWarningFormat="true"
+      WorkingDirectory="$(BuildRepositoryRoot)" />
+
+    <ItemGroup>
+      <ArtifactsToCopy Include="$(RepositoryArtifactsDirectory)*" />
+    </ItemGroup>
+
+    <Copy
+      SourceFiles="@(ArtifactsToCopy)"
+      DestinationFolder="$(BuildDir)" />
+
+    <MSBuild
+      Projects="$(MSBuildThisFileDirectory)PublishPackages.targets"
+      Condition="'$(PublishPackages)'=='true'"
+      Properties="
+        PackagesDirectory=$(RepositoryArtifactsDirectory);
+        DotNetPath=$(DotNetPath);
+        NuGetFeed=$(NuGetPublishVolatileFeed);
+        APIKey=$(APIKey)" />
+
+    <Message Text="============ Done building $(RepositoryToBuild) ============" Importance="High" />
+  </Target>
+
+  <Target Name="_PinVersions">
```

Comment:
Summary of what we're doing here: prior to building a repository:

1. We look at the restore graph spec to determine packages with floating versions
2. Look at `artifacts/build/*.nupkg and `.deps/build/*.nupkg` to determine if we have a package that has the same `{Major.Minor.Build}` number as the floated version
3. Pin the build if we found something in either of the two packages. The resulting file looks like this:

```xml
<Project>
  <ItemGroup>
    <PackageReference Update="Microsoft.AspNetCore.Http.Abstractions" Version="1.2.0-preview1-t0041a40df" Condition="'$(TargetFramework)'=='net451'" />
    <PackageReference Update="Microsoft.AspNetCore.Http.Extensions" Version="1.2.0-preview1-t0041a40df" Condition="'$(TargetFramework)'=='net451'" />
    <PackageReference Update="Microsoft.AspNetCore.WebUtilities" Version="1.2.0-preview1-t0041a40df" Condition="'$(TargetFramework)'=='net451'" />
    <PackageReference Update="Microsoft.AspNetCore.Http.Abstractions" Version="1.2.0-preview1-t0041a40df" Condition="'$(TargetFramework)'=='netstandard1.3'" />
    <PackageReference Update="Microsoft.AspNetCore.Http.Extensions" Version="1.2.0-preview1-t0041a40df" Condition="'$(TargetFramework)'=='netstandard1.3'" />
    <PackageReference Update="Microsoft.AspNetCore.WebUtilities" Version="1.2.0-preview1-t0041a40df" Condition="'$(TargetFramework)'=='netstandard1.3'" />
  </ItemGroup>
</Project>
```
4. We drop this file in the project's obj directory with the right naming so that the sdk's auto include picks it up.

This removes the need for us to juggle around with the feeds to produce a coherent build. It also works around issues like the one Kestrel hit the other day where a higher version package (that we hadn't mirrored) ended up on NuGet.